File: packages/agent/src/tools/skill.ts
import { tool } from 'ai'
import { z } from 'zod'

interface SkillToolParams {
  useSkill: (skill: string) => void
}

export const getSkillTools = ({ useSkill }: SkillToolParams) => {
  const useSkillTool = tool({
    description: 'Use a skill if you think it is relevant to the current task',
    inputSchema: z.object({
      skillName: z.string().describe('The name of the skill to use'),
      reason: z.string().describe('The reason why you think this skill is relevant to the current task'),
    }),
    execute: async ({ skillName, reason }) => {
      useSkill(skillName)
      return {
        success: true,
        skillName,
        reason,
      }
    },
  })

  return {
    'use_skill': useSkillTool,
  }
}

File: internal/agent/tools/skill.go
package tools

import (
	"context"
	"errors"
	"log/slog"

	sdk "github.com/memohai/twilight-ai/sdk"
)

type SkillProvider struct {
	logger *slog.Logger
}

func NewSkillProvider(log *slog.Logger) *SkillProvider {
	if log == nil {
		log = slog.Default()
	}
	return &SkillProvider{logger: log.With(slog.String("tool", "skill"))}
}

func (*SkillProvider) Tools(_ context.Context, session SessionContext) ([]sdk.Tool, error) {
	if session.IsSubagent {
		return nil, nil
	}
	return []sdk.Tool{
		{
			Name:        "use_skill",
			Description: "Use a skill if you think it is relevant to the current task",
			Parameters: map[string]any{
				"type": "object",
				"properties": map[string]any{
					"skillName": map[string]any{
						"type":        "string",
						"description": "The name of the skill to use",
					},
					"reason": map[string]any{
						"type":        "string",
						"description": "The reason why you think this skill is relevant to the current task",
					},
				},
				"required": []string{"skillName", "reason"},
			},
			Execute: func(_ *sdk.ToolExecContext, input any) (any, error) {
				args := inputAsMap(input)
				skillName := StringArg(args, "skillName")
				reason := StringArg(args, "reason")
				if skillName == "" {
					return nil, errors.New("skillName is required")
				}
				return map[string]any{
					"success":   true,
					"skillName": skillName,
					"reason":    reason,
				}, nil
			},
		},
	}, nil
}
